Add docker setup using WP-ENV

moon0326 opened the pull request:
This PR adds WooCommerce Admin Docker setup using [WP-ENV](https://github.com/WordPress/gutenberg/tree/master/packages/env).

~~I really like how simple WP-ENV is, but there is one limitation. The current version of WP-ENV does not support creating a custom Dockerfile. Due to the limitation, we need to install node and composer manually. There is a [draft PR ](https://github.com/WordPress/gutenberg/pull/26047)to improve the limitation. You need to run `npm run docker-dev -- install` every time you stop and start the container, but this won't take a minute.~~

### Detailed test instructions:

README can be found [here](https://github.com/woocommerce/woocommerce-admin/blob/add/docker-local-env-with-wp-env/docker/wc-admin-docker-dev/README.md), but you can start with just a single command if you already have `WP-ENV` in your system.

1. Checkout this branch.
2. Run `wp-env start`

No changelog required

Commit:
Add docker setup using WP-ENV

moon0326 commented:
Hey @samueljseay  and @adrianduffell,

Could you review this PR when you get a chance? This is a follow-up PR using WP-ENV as suggested.

Thank you 🙇

samueljseay commented on docker/wc-admin-docker-dev/README.md:
```diff
@@ -0,0 +1,111 @@
+
+You can create `.wp-env.override.json` file and override the default configuration values.
+
+You can find more about `.wp-env.override.json` configuration [here](https://github.com/WordPress/gutenberg/tree/master/packages/env#wp-envoverridejson).
```
Its great that we link to the docs, but I wonder if a couple of things from the docs deserve highlighting here to help newbies out. Two that come to mind:

1. Adding other local plugins to your override so they can be devved on to. Great if you have a local copy of woocommerce and woocommerce-admin that you want to dev on together

2. Setting up xdebug. This is something that can be a bit daunting but its really important to have access to a working xdebug config. It could just be as simple as explicitly linking this section: https://github.com/WordPress/gutenberg/tree/master/packages/env#using-xdebug

moon0326 replied:
Good idea 👍 I will update the doc.

samueljseay reviewed (commented):
A few comments from me, my main confusion is around installing things like npm and composer so that you can do things like npm install, because I'm 99% sure that wp-env uses bind mounts like most other wordpress docker env setups which means you shouldn't need that.

moon0326 commented on docker/wc-admin-docker-dev/README.md:
```diff
@@ -0,0 +1,111 @@
+
+### Using Docker for Both Backend and Frontend
+
+Run `npm run docker-dev -- setup`.
```
You're correct. WP-ENV mounts the plugin directory.

You can also run the composer and npm locally, but not everyone has a composer available locally. Since this docker setup is intended to work as it is without anything from the user's local environment, I think we should have those commands for those without npm and composer.

moon0326 replied:
We can also remove NPM related command completely and ask users to run them locally, but I think composer command should stay as it's harder to setup (Installing and switching versions) PHP environment than node.

samueljseay replied:
@moon0326 I don't think this makes sense. People should have those available locally and running them in the container is a layer of indirection that makes no sense.

samueljseay replied:
As another argument if you *really* wanted to install composer and npm for them, then you *still* don't need to do it in the container because its a bind mount, you'd just have a script to do it for them locally.

moon0326 replied:
@samueljseay The main goal of having a Docker setup is to have the exact same environment (Linux distro, PHP version, PHP settings, and so on) and minimize local setup in my opinion. It's pretty easy to install Node on Mac OS, Linux, and Windows. The problem is PHP and Composer.

Mac OS comes with PHP installed by default, but Linux and Windows don't. Installing PHP can be pretty tricky depending on local configuration.

With that said, I totally understand your concern and point. If we remove NPM and Composer installation from the container, things get much simpler. I would be happier, to be honest. 

I'm going to remove them and update the doc. If anyone complains about them later, I think we can add them again.

Thank you for the input! Updates are coming very soon 👍

adrianduffell replied:
> If we remove NPM and Composer installation from the container, things get much simpler. I would be happier, to be honest.

That sounds good. Keep it simple 👍 

As a wp-env user, I just use npm and composer locally. However, wp-env does have a composer service out of the box. I think that could be used instead if folks don't want a local composer install.

https://github.com/WordPress/gutenberg/blob/8ab2453b7afffcee9f383d668399778597c4e181/packages/env/lib/build-docker-compose-config.js#L214-L216

moon0326 replied:
@adrianduffell Thank you!! Good find!!! 👍

moon0326 commented on docker/wc-admin-docker-dev/README.md:
```diff
@@ -0,0 +1,111 @@
+`npm run docker-dev -- install`
+
+
+### Using Docker for Backend Only
```
Good idea, I will re-write it.

adrianduffell commented on .wp-env.json:
```diff
@@ -0,0 +1,13 @@
+        ".", 
+        "https://downloads.wordpress.org/plugin/woocommerce.5.0.0.zip"
+    ],
+    "config": {
```
I've found wp-env has some problems running crons. I add `"ALTERNATE_WP_CRON": true` as a workaround.

https://github.com/WordPress/gutenberg/issues/20569#issuecomment-689203786

moon0326 replied:
Thank you! I will add it.

Commit:
Add ALTERNATE_WP_CRON config to avoid cron issues

Commit:
Add changelog

moon0326 commented:
@samueljseay @adrianduffell 

Thank you for all your inputs 👍 I've pushed the requested changes.

1. Removed all `npm run docker-dev` commands. We only have `npm run wp-env-mysql-port` to get the MySQL port. I'm going to see if I can open a PR in WP-ENV to add a command to output MySQL port. 
2. Updated README.
3. Added `ALTERNATE_WP_CRON` as suggested by Adrian.

adrianduffell commented on docker/wc-admin-wp-env/README.md:
```diff
@@ -0,0 +1,113 @@
+	"plugins": [
+        ".", 
+        "https://downloads.wordpress.org/plugin/woocommerce.5.0.0.zip",
+        "../wp-crontrol"
```
I wonder if we should swap these examples around, so that there is a URL to wp-crontrol, and woocommerce is loaded from a local path. I think that would match a typical setup.

moon0326 replied:
@adrianduffell Good idea 👍 I will update it.

moon0326 replied:
@adrianduffell Updated in https://github.com/woocommerce/woocommerce-admin/pull/6448/commits/00c7d6fc9f955ca274d3e1fccde4e6c1ad461fbe

adrianduffell commented:
> Thank you for all your inputs 👍 I've pushed the requested changes.

Great work here @moon0326, I love how simple this has become. I haven't had a chance to test this yet -- Tuesday is probably the earliest I can. But don't wait for me if it's ready to merge before then 😄

adrianduffell commented:
It looks like this needs a merge with `main` to get the most recent CI checks.

moon0326 commented:
> > Thank you for all your inputs 👍 I've pushed the requested changes.
> 
> Great work here @moon0326, I love how simple this has become. I haven't had a chance to test this yet -- Tuesday is probably the earliest I can. But don't wait for me if it's ready to merge before then 😄

Thank you Adrian! I'm going to update README with a few HOWTOs and rebase the main.

Commit:
Update README

* Added HOWTOs section

* Updated .wp-env.override.json examples

adrianduffell reviewed (approved):
This tested well and LGTM 🚀 

`wp-env start` started the site as expected!

 `npm run wp-env-mysql-port` script got the correct port for me. Submitting this script upstream sounds like a good plan 👍 

Nice work on the documentation too 🎉
